spec/unit/matrixrtc/MembershipManager.spec.ts

```diff
@@ -0,0 +1,611 @@
+        });
+    });
+}
+
+function createAsyncHandle(method: any) {
+    const handle: { resolve?: (...args: unknown[]) => void; reject?: (...args: any[]) => void } = {};
+    (method as Mock).mockImplementation(() => {
+        return new Promise((resolve, reject) => {
+            handle.reject = reject;
+            handle.resolve = resolve;
+        });
+    });
+    return handle;
+}
+
+/**
+ * Tests different MembershipManager implementations. Some tests don't apply to `LegacyMembershipManager`
+ * use !FailsForLegacy to skip those. See: testEnvironment for more details.
+ */
+describe("MembershipManager", () => {
+    describe.each([
+        { TestMembershipManager: LegacyMembershipManager, description: "LegacyMembershipManager" },
+        // Here we will add the new implementation of the MembershipManager.
+        // It is not yet tested since it would currently fail all tests. Adding the MembershipManger looks like this:
+        // { TestMembershipManager: MembershipManager, description: "MembershipManager" },
+    ])("$description", ({ TestMembershipManager }) => {
+        let client: MockClient;
+        let room: Room;
+        const focusActive: LivekitFocusActive = {
+            focus_selection: "oldest_membership",
+            type: "livekit",
+        };
+        const focus: Focus = {
+            type: "livekit",
+            livekit_service_url: "https://active.url",
+            livekit_alias: "!active:active.url",
+        };
+
+        beforeEach(() => {
+            // default to fake timers
+            jest.useFakeTimers();
+            client = makeMockClient("@alice:example.org", "AAAAAAA");
+            room = makeMockRoom(membershipTemplate);
+            // provide a default mock that is like the default "non error" server behaviour.
+            (client._unstable_sendDelayedStateEvent as Mock).mockReturnValue({ delay_id: "id" });
+        });
+
+        afterEach(() => {
+            jest.useRealTimers();
+            // no need to clean up mocks since we will recreate the client
+        });
+
+        describe("isJoined()", () => {
+            it("defaults to false", () => {
+                const manager = new TestMembershipManager({}, room, client, () => undefined);
+                expect(manager.isJoined()).toEqual(false);
+            });
+
+            it("returns true after join()", async () => {
+                const manager = new TestMembershipManager({}, room, client, () => undefined);
+                manager.join([]);
+                expect(manager.isJoined()).toEqual(true);
+            });
+        });
+
+        describe("join()", () => {
+            describe("sends a membership event", () => {
+                it("sends a membership event and schedules delayed leave when joining a call", async () => {
+                    // Spys/Mocks
+
+                    // eslint-disable-next-line camelcase
+                    const _unstable_updateDelayedEventHandle = createAsyncHandle(
```

Comment:
Not sure I understand the reason for the weird name here

Reply:
its `fn_name`+`Handle`. What do you think would be better? just `updateDelayedEventHandle`?